File: docs-examples/example-java/src/test/java/io/micronaut/rabbitmq/docs/parameters/BindingSpec.java
package io.micronaut.rabbitmq.docs.parameters;

import io.micronaut.context.annotation.Property;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.support.TestPropertyProvider;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import io.micronaut.rabbitmq.testcontainers.RabbitMQ;
import java.util.Map;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

@MicronautTest
@Property(name = "spec.name", value = "BindingSpec")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class BindingSpec implements TestPropertyProvider {
    @Override
    public Map<String, String> getProperties() {
        return RabbitMQ.getProperties();
    }

    @Test
    void testDynamicBinding(ProductClient productClient, ProductListener productListener) {
// tag::producer[]
        productClient.send("message body".getBytes());
        productClient.send("product", "message body2".getBytes());
// end::producer[]


        await().atMost(60, SECONDS).until(() ->
                productListener.messageLengths.size() == 2 &&
                productListener.messageLengths.contains(12) &&
                productListener.messageLengths.contains(13)
        );
    }
}

File: docs-examples/example-python/src/test/python/micronaut/rabbitmq/docs/parameters/BindingSpec.py
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.rabbitmq.docs.Await import Await
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .ProductClient import ProductClient
from .ProductListener import ProductListener


@MicronautTest(environments=["rabbitmq"])
@Property(name="spec.name", value="BindingSpec")
class BindingSpec:
    product_client: Annotated[ProductClient, Inject]
    product_listener: Annotated[ProductListener, Inject]

    @Test
    def test_dynamic_binding(self) -> None:
# tag::producer[]
        self.product_client.send(b"message body")
        self.product_client.send_to("product", b"message body2")
# end::producer[]

        message_lengths = self.product_listener.message_lengths
        Await.until(lambda: len(message_lengths) == 2 and 12 in message_lengths and 13 in message_lengths)
